components/List/src/MDCSelfSizingStereoCell.h

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2018-present the Material Components for iOS authors. All Rights Reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#import <UIKit/UIKit.h>
+#import "MDCBaseCell.h"
+
+@interface MDCSelfSizingStereoCell : MDCBaseCell
+
+/**
+ The UIImageView responsible for displaying the leading image.
+ */
+@property (nonatomic, strong, readonly) UIImageView *leadingImageView;
+
+/**
+ The UIImageView responsible for displaying the trailing image.
+ */
+@property (nonatomic, strong, readonly) UIImageView *trailingImageView;
+
+/**
+ The UILabel responsible for displaying the title text.
+ */
+@property (nonatomic, strong, readonly) UILabel *titleLabel;
+
+/**
+ The UILabel responsible for displaying the detail text.
+ */
+@property (nonatomic, strong, readonly) UILabel *detailLabel;
+
+/**
+ Sepcifies the width of the cell. The layout of the views and the height of the cell are determined
+ by the value of this property.
+ */
+@property (nonatomic, assign) CGFloat cellWidth;
```

Comment:
If possible can we remove this?

Reply:
@jverkoey, I know this property has been fairly controversial, but I can't think of a good way to get rid of it. I just tried again for about an hour haha. At one point it was private... I was setting it in `-systemLayoutSizeFittingSize:` to the value of the `width` property of the `targetSize` parameter of that method and then calculating the subview frames and height from there. This worked because when `estimatedItemSize` is set on a flow layout the system calls `-systemLayoutSizeFittingSize:` very shortly before the cell is displayed with the `estimatedItemSize` as the parameter. However, I realized that if anyone else called that method it would have a side effect of transforming the layout of the cell, which didn't seem like something a method that's just supposed to return the size should do. I decided it was better to have the client set the width. With self-sizing cells I think this is fairly normal. It seems to be in many, if not all, of the blog posts I've seen about self-sizing UICollectionView (not UITableView) cells. This same property also exists in Yarden's list cells he made before I joined.

Reply:
Have you seen this example code: https://developer.apple.com/documentation/uikit/uifont/creating_self_sizing_table_view_cells ? It looks like they're able to make the cells calculate their height automatically without providing a width, but it also looks like they're not overriding `-systemLayoutSizeFittingSize:`. Is this a simpler example than what we're trying to achieve and therefor doesn't apply?

Reply:
@jverkoey , I think the situation in that example is a little different because it's a TableView, where width is less variable, and it's using Auto Layout. 

How about this--maybe we could maybe take advantage of `-preferredLayoutAttributesFittingAttributes:`? We could put the side effect (setting `cellWidth`) there. The documentation on this method actually specifies that its purpose is to "adjust the size values to accommodate changes made by a self-sizing cell." I think that mainly refers to changes in the `layoutAttributes`, but if the collection view tells us it has a given width in mind that might be enough to justify calculating a cell layout based on that width. So we could do something like this:

```
-(CGSize)systemLayoutSizeFittingSize:(CGSize)targetSize {
  if (!self.hasCalculatedLayout) {
    [self calculateLayout];
  }
  return CGSizeMake(self.cellWidth, self.calculatedHeight);
}

-(UICollectionViewLayoutAttributes *)preferredLayoutAttributesFittingAttributes:
    (UICollectionViewLayoutAttributes *)layoutAttributes {
  self.cellWidth = layoutAttributes.size.width;
  UICollectionViewLayoutAttributes *attributes =
      [super preferredLayoutAttributesFittingAttributes:layoutAttributes];
  return attributes;
}
```
Note that `-systemLayoutSizeFittingSize:` gets called some time after `-preferredLayoutAttributesFittingAttributes:` is entered and some time before it returns.

I think the public cellWidth is a little safer/less likely to lead to surprises, but I think this would probably work for like 99% of cases? I think only the collection view is ever supposed to call this method, whereas I can imagine a client calling `-systemLayoutSizeFittingSize:`, although that too is unlikely. Honestly with the way it's built neither method should really need to be called :/